# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

load_category <- function(anno_start, anno_end, category_file, SNP = "SNP", CHR = "CHR", BP = "BP") {
    .Call('SSREM_load_category', PACKAGE = 'SSREM', anno_start, anno_end, category_file, SNP, CHR, BP)
}

load_summary <- function(summary_file, beta = "b", sd = "SE", N = "N", Allele1 = "Allele1", Allele2 = "Allele2", SNP = "SNP", CHR = "CHR", BP = "BP") {
    .Call('SSREM_load_summary', PACKAGE = 'SSREM', summary_file, beta, sd, N, Allele1, Allele2, SNP, CHR, BP)
}

#' @title
#' SSREM_More
#' @description
#' Continue SSREM Gibbs Training after previous gibbs procedure
#' @param more number of iterations
#' @param num_thread number of threads
#'
#' @param cache file name of the cache for the gibbs results to recover the model
#'
#' @return List of gibbs results, the same as the output of SSREM function
SSREM_More <- function(more, num_thread = -1L, cache = "ssrem.cache", logfile = "screen") {
    .Call('SSREM_SSREM_More', PACKAGE = 'SSREM', more, num_thread, cache, logfile)
}

#' @title
#' SSREM
#' @description
#' SSREM Gibbs Training
#'
#' @param geno_file  genotype data file of plink data format
#'
#' @param block_file  block file for the block boundary of snps
#'
#' @param summaries  input of summary statistics, data frames with column names "snp,A1,A2,b,se,N"
#'
#' @param annotations input of category information, data frames with snp column and category informations
#'
#' @param beta identifier for beta, default value is b
#'
#' @param sd identifier for se, default value is se
#'
#' @param N identifier for N(sample size), default value is N
#'
#' @param Allele1 identifier for allele 1, default value is A1
#'
#' @param Allele2 identifier for allele 2, default value is A2
#'
#' @param SNP identifier for SNP, default value is SNP
#'
#' @param CHR identifier for Chromsome, default value is CHR
#'
#' @param BP identifier for Base Pair, default value is BP
#'
#' @param num_thread number of threads, default value is 1
#'
#' @param burnin number of burnin for gibbs, default value is 100
#'
#' @param numit number of iteration for gibbs after burnin, default value is 2000
#'
#' @param thin value of interval for recording gibbs results, default value is 10
#'
#' @param logfile indicate the log file name, default value is "screen" for displaying the output information on the screen
#'
#' @param cache file name of the cache for the gibbs results to recover the model
#'
#' @return List of gibbs results, refer to the end of the function for the details
SSREM <- function(geno_file, block_file, summaries, annotations, beta = "b", sd = "se", N = "N", Allele1 = "A1", Allele2 = "A2", SNP = "SNP", CHR = "CHR", BP = "BP", num_thread = 1L, burnin = 100L, numit = 2000L, thin = 10L, logfile = "screen", cache = "ssrem.cache") {
    .Call('SSREM_SSREM', PACKAGE = 'SSREM', geno_file, block_file, summaries, annotations, beta, sd, N, Allele1, Allele2, SNP, CHR, BP, num_thread, burnin, numit, thin, logfile, cache)
}

